File: folder_size.py
# folder_size.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import argparse
from pathlib import Path
from collections import defaultdict

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

def human_readable(size, decimal_places=1):
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024.0:
            return f"{size:{decimal_places+1}.{decimal_places}f} {unit}"
        size /= 1024.0
    return f"{size:.1f} PB"

def get_folder_size(path, recursive=True, depth=0, max_depth=None, exclude_hidden=False, verbose=False):
    """Рекурсивно вычисляет размер папки и возвращает словарь {путь: размер}."""
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"Path not found: {path}")
    if not path.is_dir():
        raise NotADirectoryError(f"Not a directory: {path}")

    result = {}
    total_size = 0

    try:
        items = list(path.iterdir())
    except PermissionError:
        if verbose:
            print(colorize(f"Permission denied: {path}", 'red'))
        return {str(path): 0}

    for item in items:
        if exclude_hidden and item.name.startswith('.'):
            continue
        if item.is_file():
            try:
                size = item.stat().st_size
                total_size += size
                if verbose:
                    print(f"  {item}: {human_readable(size)}")
            except OSError:
                pass
        elif item.is_dir() and recursive and (max_depth is None or depth < max_depth):
            sub_result = get_folder_size(item, recursive, depth+1, max_depth, exclude_hidden, verbose)
            if sub_result:
                sub_size = sum(sub_result.values())
                total_size += sub_size
                result.update(sub_result)
        elif item.is_dir() and not recursive:
            # Если не рекурсивно, просто пропускаем папки
            pass

    result[str(path)] = total_size
    return result

def main():
    parser = argparse.ArgumentParser(description="FolderSize – измерение размера папок")
    parser.add_argument('path', nargs='?', default='.', help='Путь к папке (по умолчанию текущая)')
    parser.add_argument('-r', '--recursive', action='store_true', default=True, help='Рекурсивно (включено)')
    parser.add_argument('-h', '--human-readable', action='store_true', help='Выводить размер в удобном формате')
    parser.add_argument('-s', '--sort', action='store_true', help='Сортировать по размеру (убывание)')
    parser.add_argument('-t', '--top', type=int, help='Показать только N самых больших элементов')
    parser.add_argument('-d', '--depth', type=int, help='Максимальная глубина рекурсии (0 – только текущая)')
    parser.add_argument('--exclude-hidden', action='store_true', help='Игнорировать скрытые файлы и папки')
    parser.add_argument('-v', '--verbose', action='store_true', help='Подробный вывод')
    args = parser.parse_args()

    try:
        size_map = get_folder_size(args.path, args.recursive, 0, args.depth,
                                   args.exclude_hidden, args.verbose)
    except Exception as e:
        sys.exit(colorize(f"Ошибка: {e}", 'red'))

    if not size_map:
        print(colorize("Нет данных.", 'yellow'))
        return

    # Преобразуем в список (путь, размер)
    items = list(size_map.items())
    if args.sort:
        items.sort(key=lambda x: x[1], reverse=True)
    if args.top:
        items = items[:args.top]

    # Определяем максимальную длину пути для выравнивания
    max_path_len = max(len(p) for p, _ in items) if items else 0

    for path, size in items:
        if args.human_readable:
            size_str = human_readable(size)
        else:
            size_str = f"{size:,} B"
        # Цвет для размера
        if size < 1024 * 1024:
            color = 'green'
        elif size < 1024 * 1024 * 1024:
            color = 'yellow'
        else:
            color = 'red'
        print(f"{colorize(size_str.rjust(12), color)}  {path}")

if __name__ == '__main__':
    main()

File: folder_size.go
// folder_size.go
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

func humanReadable(size int64) string {
	units := []string{"B", "KB", "MB", "GB", "TB"}
	s := float64(size)
	for _, u := range units {
		if s < 1024.0 {
			return fmt.Sprintf("%.1f %s", s, u)
		}
		s /= 1024.0
	}
	return fmt.Sprintf("%.1f PB", s)
}

type Item struct {
	Path string
	Size int64
}

func getFolderSize(path string, recursive bool, depth int, maxDepth int, excludeHidden bool, verbose bool) ([]Item, error) {
	info, err := os.Stat(path)
	if err != nil {
		return nil, err
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("not a directory: %s", path)
	}

	var result []Item
	var totalSize int64

	entries, err := os.ReadDir(path)
	if err != nil {
		if verbose {
			fmt.Println(colorize("Permission denied: "+path, red))
		}
		return []Item{{Path: path, Size: 0}}, nil
	}

	for _, entry := range entries {
		if excludeHidden && strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		fullPath := filepath.Join(path, entry.Name())
		info, err := entry.Info()
		if err != nil {
			continue
		}
		if entry.IsDir() {
			if recursive && (maxDepth == 0 || depth < maxDepth) {
				subItems, err := getFolderSize(fullPath, recursive, depth+1, maxDepth, excludeHidden, verbose)
				if err != nil {
					if verbose {
						fmt.Println(colorize("Error reading "+fullPath+": "+err.Error(), red))
					}
					continue
				}
				var subTotal int64
				for _, item := range subItems {
					subTotal += item.Size
				}
				totalSize += subTotal
				result = append(result, subItems...)
			}
		} else {
			size := info.Size()
			totalSize += size
			if verbose {
				fmt.Printf("  %s: %s\n", fullPath, humanReadable(size))
			}
		}
	}
	result = append(result, Item{Path: path, Size: totalSize})
	return result, nil
}

func main() {
	var (
		path          string
		recursive     bool
		human         bool
		sortFlag      bool
		top           int
		depth         int
		excludeHidden bool
		verbose       bool
	)
	flag.StringVar(&path, "p", ".", "Path")
	flag.BoolVar(&recursive, "r", true, "Recursive")
	flag.BoolVar(&human, "h", false, "Human-readable")
	flag.BoolVar(&sortFlag, "s", false, "Sort by size")
	flag.IntVar(&top, "t", 0, "Show top N")
	flag.IntVar(&depth, "d", 0, "Max depth (0=unlimited)")
	flag.BoolVar(&excludeHidden, "exclude-hidden", false, "Exclude hidden")
	flag.BoolVar(&verbose, "v", false, "Verbose")
	flag.Parse()
	if flag.NArg() > 0 {
		path = flag.Arg(0)
	}

	absPath, err := filepath.Abs(path)
	if err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		os.Exit(1)
	}

	items, err := getFolderSize(absPath, recursive, 0, depth, excludeHidden, verbose)
	if err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		os.Exit(1)
	}

	// Собираем итоговые размеры папок в map
	sizeMap := make(map[string]int64)
	for _, item := range items {
		sizeMap[item.Path] = item.Size
	}

	// Преобразуем в слайс для сортировки
	var result []Item
	for p, s := range sizeMap {
		result = append(result, Item{Path: p, Size: s})
	}
	if sortFlag {
		sort.Slice(result, func(i, j int) bool {
			return result[i].Size > result[j].Size
		})
	}
	if top > 0 && top < len(result) {
		result = result[:top]
	}

	maxLen := 0
	for _, item := range result {
		if len(item.Path) > maxLen {
			maxLen = len(item.Path)
		}
	}
	for _, item := range result {
		var sizeStr string
		if human {
			sizeStr = humanReadable(item.Size)
		} else {
			sizeStr = fmt.Sprintf("%d B", item.Size)
		}
		color := green
		if item.Size > 1024*1024*1024 {
			color = red
		} else if item.Size > 1024*1024 {
			color = yellow
		}
		fmt.Printf("%s  %s\n", colorize(fmt.Sprintf("%12s", sizeStr), color), item.Path)
	}
}
